More control over flow ( interrupt command JIT  + choose commands to ignore/stop on + stop when looping )

eyalk11 opened the pull request:
<!-- ⚠️ At the moment any non-essential commands are not being merged.
If you want to add non-essential commands to Auto-GPT, please create a plugin instead.
We are expecting to ship plugin support within the week (PR #757).
Resources:
* https://github.com/Significant-Gravitas/Auto-GPT-Plugin-Template
-->

<!-- 📢 Announcement
We've recently noticed an increase in pull requests focusing on combining multiple changes. While the intentions behind these PRs are appreciated, it's essential to maintain a clean and manageable git history. To ensure the quality of our repository, we kindly ask you to adhere to the following guidelines when submitting PRs:

Focus on a single, specific change.
Do not include any unrelated or "extra" modifications.
Provide clear documentation and explanations of the changes made.
Ensure diffs are limited to the intended lines — no applying preferred formatting styles or line endings (unless that's what the PR is about).
For guidance on committing only the specific lines you have changed, refer to this helpful video: https://youtu.be/8-hSNHHbiZg

Check out our [wiki page on Contributing](https://github.com/Significant-Gravitas/Auto-GPT/wiki/Contributing)

By following these guidelines, your PRs are more likely to be merged quickly after testing, as long as they align with the project's overall direction. -->

### Background
<!-- Provide a concise overview of the rationale behind this change. Include relevant context, prior discussions, or links to related issues. Ensure that the change aligns with the project's overall direction. -->
Now it is quite challenging to control the flow. You would need to assume the next N authorised commands are OK, beforehand. If it goes astray, you have nothing to do, but to wait after the batch. 
This commit make it possible to interrupt just in time and define commands that are more and less relevant with regard to batch.
 
### Changes
<!-- Describe the specific, focused change made in this pull request. Detail the modifications clearly and avoid any unrelated or "extra" changes. -->
1. Added` config.commands_to_stop` and `config.commands_to_ignore`.

`config.commands_to_stop` would stop after the command anyway.  For example, you might want to stop on task completion, anyway. 

 `config.commands_to_ignore` are commands that authorized automatically and don't increase the counter.

2.  Changed spinner class to allow for interruption

  *  `<space>` in the middle of thinking means soft-interrupt complete execution and ask for user input after execution).  
  * `<q>` means abort - exit task immediately and ask for user input.

  Have two implementation for linux and windows.

  Introduced a function `async_task_and_spin`  to wait for either task completion or interrupt. 

3. Stop if already executed a command

  To do so, it keeps all previously executed commands in set. 

### Documentation
<!-- Explain how your changes are documented, such as in-code comments or external documentation. Ensure that the documentation is clear, concise, and easy to understand. -->
In code 

### Test Plan
<!-- Describe how you tested this functionality. Include steps to reproduce, relevant test cases, and any other pertinent information. -->
I just checked the code on my own, saw that both soft and hard interrupts working.
commands_to_stop and to_ignore seems to be working.
There was an error in test_memory_trimmed_from_context that seem unrelated.
```
    memory_found = memory.get_relevant("Important Information", 5)
    assert memory_found[0] == expected_permanent_memory
```
Want to see if it fails on workflow.
### PR Quality Checklist
- [ ] My pull request is atomic and focuses on a single change. (Arguably, you could split this into adding ignore/to stop and adding spinner. But it is part of whole. )
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes <!-- Submit these as separate Pull Requests, they are the easiest to merge! -->

<!-- If you haven't added tests, please explain why. If you have, check the appropriate box. If you've ensured your PR is atomic and well-documented, check the corresponding boxes. -->
I didn't add tests because of lack of time. A bit of help is wanted, if you think testing is required for it. 

This is how soft interrupt looks like. It keeps thinking afterwards. 
![image](https://user-images.githubusercontent.com/72234965/236655300-fd7a5eeb-b405-41eb-82b1-c225c9386ef4.png)

eyalk11 commented:
There is a tricky thing here. It needs termios and tty for linux and keyboard for windows. as termios failed to install on windows , I couldn't include this in requirments.

Pwuts commented:
This looks interesting, and like a significant potential UX improvement

lc0rp commented:
@eyalk11 We're prepping for release v0.4.3. Please resolve conflicts and stand by as we merge.

eyalk11 commented:
OK, I am working on this. it is a bit hard to merge.  (merge_int branch in my fork).

eyalk11 commented:
(There are still some merging issues, will solve)

eyalk11 commented:
I removed the check for reoccurring commands (8139493dd98bab7b474cbbf89b24d6f19275d908) . And added comment in https://github.com/Significant-Gravitas/Auto-GPT/issues/3668 . I will open a new PR for it. 

Spinner for windows nt requires keyboard, currently not in requirements (will fail for linux).

eyalk11 commented:
There is a bug. The bug is that there are leftover executation threads after async_task_and_spin is called because 'pool.shutdown(wait=True)' isn't called. 
Simply doing.
```
loop.call_later(0.1, lambda: pool.shutdown(wait=True)) 
```
in finally won't solve it, because `asyncio.run` will wait until it is called.

That is why, as long as we don't run natively with asyncio, we need another thread to handle it. Otherwise, we could return a future and execute it opportunistically.  My suggested solution is like this: 
https://stackoverflow.com/questions/32059732/send-asyncio-tasks-to-loop-running-in-other-thread

I am open to suggestions. But canceling task without waiting for it to be canceled is problematic.

Pwuts reviewed (commented):
Interrupting continuous execution is currently possible with `Ctrl`+`C`, as implemented by #4230. Is there still more to be done on this front?

The "commands to ignore" and "commands to stop on" are a really good idea, as well as hard/soft interrupts! If you could submit these in separate pull requests, that would be grand (makes it a lot easier to review). :)

eyalk11 commented on autogpt/spinner.py:
```diff
@@ -13,44 +38,96 @@ def __init__(
         return self
 
-    def __exit__(self, exc_type, exc_value, exc_traceback) -> None:
+    def __exit__(self, exc_type, exc_value, exc_traceback) -> None:  # type: ignore
```
I was too lazy to check the exact signature. Also not my line, yet prevented merge.

eyalk11 replied:
fixed __exit__

eyalk11 commented:
@Pwuts and @lc0rp I undo refactoring per your request. Consider including this in v0.4.4 . Created branch https://github.com/eyalk11/Auto-GPT/tree/interrupt_commands_with_refactoring that I can PR if you want.

dschonholtz commented:
# AI-Maintainer Review for PR - More control over flow (interrupt command JIT + choose commands to ignore/stop on + stop when looping)

## Title and Description :+1:

<summary>The Title and Description are clear, concise, and helpful</summary>
<details>
The title and description of the pull request are clear and concise. They effectively communicate the purpose of the changes, which is to provide more control over the flow in the Auto-GPT project. The description provides a detailed explanation of the changes and their expected behavior.
</details>

## Scope of Changes :+1:

<summary>The changes are narrowly focused</summary>
<details>
The changes in the pull request are narrowly focused on addressing a specific issue related to controlling the flow in the Auto-GPT project. The modifications all contribute to the same goal of providing more control over the flow, and it does not appear that the author is trying to resolve multiple unrelated issues simultaneously.
</details>

## Testing :warning:

<summary>The testing approach could be more comprehensive</summary>
<details>
The author has tested the changes and provided a description of the testing process. However, the testing approach could be more comprehensive. It would be beneficial to include more detailed test cases and potentially automated tests to ensure the new features work as expected in various scenarios.
</details>

## Documentation :warning:

<summary>Some functions, classes, or methods lack docstrings</summary>
<details>
The following functions, classes, or methods lack docstrings:

- `async_task_and_spin` in the `Agent` class
- `start_interaction_loop` in the `Agent` class
- `_resolve_pathlike_command_args` in the `Agent` class
- `spin` in the `Spinner` class
- `update_message` in the `Spinner` class

These should be updated to include docstrings describing their behavior, arguments, and return values.
</details>

## Suggested Changes

- Add docstrings to the functions, classes, or methods mentioned above. The docstrings should describe the behavior, arguments, and return values of each. This will improve code readability and maintainability.
- Consider enhancing the testing approach to include more detailed test cases and potentially automated tests. This will help ensure the new features work as expected in various scenarios and catch any potential issues early.

## Potential Issues

- The `async_task_and_spin` function in the `Agent` class uses asyncio and threading together, which can lead to complex and hard-to-debug issues. Consider refactoring this to use only asyncio or only threading, if possible.
- The `Spinner` class now has an `interruptable` attribute, but it's not clear how this is used if the spinner is not interruptable. Consider adding comments or documentation to clarify this.

eyalk11 commented:
I did restructuring (undoing refactoring) and merged master.

eyalk11 commented:
There is a bug in uploading test results 
```
Run actions/upload-artifact@v3
With the provided path, there will be 47 files uploaded
Starting artifact upload
For more detailed logs during the artifact upload process, enable step-debugging: https://docs.github.com/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging#enabling-step-debug-logging
Artifact name is valid!
Error: Artifact path is not valid: /DEBUG/20230714_184835_<MagicMock name='mock.ai_name.__getitem__().rstrip()' id='[13](https://github.com/Significant-Gravitas/Auto-GPT/actions/runs/5557322292/jobs/10150954433?pr=3914#step:16:14)9770235010304'>/001/0_full_message_history.json. Contains the following character:  Less than <
          
Invalid characters include:  Double quote ", Colon :, Less than <, Greater than >, Vertical bar |, Asterisk *, Question mark ?, Carriage return \r, Line feed \n
          
The following characters are not allowed in files that are uploaded due to limitations with certain file systems such as NTFS. To maintain file system agnostic behavior, these characters are intentionally not allowed to prevent potential problems with downloads on different file systems.

```
That is why it fails. 

I added test for command_to_stop/ignore.

eyalk11 commented on tests/unit/test_ignore_stop_commands.py:
```diff
@@ -0,0 +1,95 @@
+    # with mock.patch("autogpt.workspace.workspace.Workspace", autospec=Workspace):
+    mocker.patch("autogpt.agents.agent.execute_command", side_effect=set_continous)
+    # For some reason config stopped working if it was a decorator
+    with patch.object(
```
As mentioned, for some bloody reason , didn't work  as decorators on the function.

eyalk11 commented on autogpt/spinner.py:
```diff
@@ -13,44 +39,101 @@ def __init__(
+            if (
+                pressed_key
+                and pressed_key == key_to_check
+                or (os.name == "nt" and keyboard.is_pressed(keyboard_key))
```
not in requirement file

eyalk11 replied:
will work if out of focus

eyalk11 commented on requirements.txt:
```diff
@@ -29,6 +29,7 @@ spacy>=3.0.0,<4.0.0
 prompt_toolkit>=3.0.38
 pydantic
+frozendict
 
```
keyboard not in req file because isn't supported on mac.

lc0rp commented:
@eyalk11 The agent was updated recently, probably the reason for the conflicts.
